rcl_yaml_param_parser: add support for binary tag to load byte array parameter

romainreignier opened the pull request:
## Description

When a parameter is declared as an array of bytes (`std::vector<uint8_t>`), passing an array of integers in the command line or in a YAML file is interpreted by the YAML parser as an array of integers and produces an error.

To workaround this, I have used a base64 encoded string with the the `!!binary` tag as shown in the [YAML specification](https://yaml.org/spec/1.2.2/#24-tags).

Example:
```
ros2 run pkg_name exe_name --ros-args -p 'my_param:=!!binary AAECAw=='
```

Note that this implementation adds a dependency on OpenSSL which might not be acceptable (and the ugly global variable hack to use the custom allocator).

We might implement the base64 decoding [directly in C](https://github.com/ros2/demos/blob/472a4796bb838906a286c8c00260169873255e59/image_tools/src/burger.cpp#L39) or use another lib like [libb64](https://libb64.sourceforge.net/).

Until this question is resolved, I put this PR as a Draft.

The sequence handling is not very clear in `parse.c` and might need some improvements.

Fixes https://github.com/ros2/ros2/issues/1436

Can be seen as part of https://github.com/ros2/rcl/issues/1026

### Is this user-facing behavior change?

Yes, the byte array parameters are now supported as ros arguments or in yaml parameters files.

Until now, it was not supported at all, has decribed in https://github.com/ros2/ros2/issues/1436.

### Did you use Generative AI?

Yes, the `base64_decode()` function has been written with the help of Mistral Le Chat.

Barry-Xu-2018 commented:
@romainreignier Thank you for this PR :+1:   
From my perspective, I prefer to use libb64.  
FYI, I will take over and complete https://github.com/ros2/rcl/pull/1032.

Barry-Xu-2018 commented:
@fujitatomoya 

In YAML files, binary data is encoded using base64. As for decoding base64, we'd like to hear your thoughts. Is it better to write our own decoder or use a library like libb64 or OpenSSL?

Barry-Xu-2018 commented:
@fujitatomoya Friendly ping

fujitatomoya commented:
related issues: https://github.com/ros2/rcl/issues/1205 and https://github.com/ros2/rcl/issues/242

fujitatomoya reviewed (commented):
i'd recommend a self-contained C implementation in `rcutils` with following reasons.

- no new external dependency
- base64 decode is trivial, easily implemented in rcutrils which already has similar code.
- OpenSSL is too heavy for this use case...

fujitatomoya commented:
@romainreignier are you still working on this?

romainreignier commented:
> @romainreignier are you still working on this?

Not currently, I have missed your previous replies.
I though @Barry-Xu-2018 wanted to take over this and I don't mind.

Barry-Xu-2018 commented:
@romainreignier   

> I though @Barry-Xu-2018 wanted to take over this and I don't mind.

Maybe I make you misunderstood. I just took over this #1032. However, #1032 does not include support for the binary tag. Once this PR is merged, full support for all tags will be available.   

I think the next step is to add base64 decode/encode APIs to rcutils first.   
Do you have time to do this? If not, I can take care of it.
And then could you update this PR with base64 decode API from rcutils? Is that okay?

romainreignier commented:
Yes please, if you can implement it in rcutils, I will update this PR.
Thanks.

Barry-Xu-2018 commented:
@romainreignier 

Base64 decode/encode APIs had been added https://github.com/ros2/rcutils/pull/533.  

About how to use, you can refer to test code
https://github.com/ros2/rcutils/blob/f32243642b1d95f2dd2ea42ef468ccb377df0e13/test/test_base64.cpp#L28-L40

Please update this PR.

romainreignier committed:
rcl_yaml_param_parser: add support for binary tag to load byte arrays parameters

This allows to pass byte array parameters using the YAML
[binary tag](https://yaml.org/spec/1.2.2/#24-tags)

Example:
```
ros2 run pkg_name exe_name --ros-args -p 'calib:=!!binary AAECAw=='
```

Address https://github.com/ros2/ros2/issues/1436

Can be seen as part of https://github.com/ros2/rcl/issues/1026

Signed-off-by: Romain Reignier <romain.reignier@exail.com>

romainreignier commented:
@Barry-Xu-2018 I have updated the PR using `rcutils_decode_base64()`.

In `parse_value()`, I explicitly do not support sequence for byte array, hence no `add_to_byte_array()` needed in `add_to_array.h` because we only support byte arrays as base64 strings.

Barry-Xu-2018 commented on rcl_yaml_param_parser/src/parse.c:
```diff
@@ -661,6 +749,49 @@ _get_bool_value(
+
+  if (NULL != byte_array.buffer) {
+    allocator.deallocate(byte_array.buffer, allocator.state);
+  }
```
Usually, rcutils_uint8_array_fini() is used to free rcutils_uint8_array_t.

Barry-Xu-2018 commented on rcl_yaml_param_parser/src/parse.c:
```diff
@@ -661,6 +749,49 @@ _get_bool_value(
+  }
+
+  memcpy(dst->values, byte_array.buffer, byte_array.buffer_length);
+  dst->size = byte_array.buffer_length;
```
Is it possible to directly use the pointer to the buffer in rcutils_uint8_array_t to avoid allocation?
```c
dst->values = byte_array.buffer;
dst->size = byte_array.buffer_length;
```
If this is done, there is no need to call rcutils_uint8_array_fini() to free rcutils_uint8_array_t.

Barry-Xu-2018 commented on rcl_yaml_param_parser/src/parse.c:
```diff
@@ -456,6 +487,63 @@ rcutils_ret_t parse_value(
+          }
+          memcpy(param_value->byte_array_value->values, src->values, src->size);
+        }
+        param_value->byte_array_value->size = src->size;
```
`rcl_byte_array_t * src = (rcl_byte_array_t *)ret_val;`  
param_value->byte_array_value is also an rcl_byte_array_t *, so there is no need to allocate it again. 

```c
param_value->byte_array_value = (rcl_byte_array_t *)ret_val;
```

Barry-Xu-2018 commented on rcl_yaml_param_parser/src/parser.c:
```diff
@@ -407,6 +407,16 @@ void rcl_yaml_node_struct_print(
               printf(": %s\n", param_var->string_value);
+            } else if (NULL != param_var->byte_array_value) {
+              printf(": ");
+              for (size_t i = 0; i < param_var->byte_array_value->size; i++) {
+                if (param_var->byte_array_value->values) {
+                  printf(
+                    "0x%02x, ",
+                    param_var->byte_array_value->values[i]);
+                }
+              }
```
There’s no need to check param_var->byte_array_value->values in every iteration of the loop.

```c
              if (param_var->byte_array_value->values) {
                for (size_t i = 0; i < param_var->byte_array_value->size; i++) {
                  printf(
                    "0x%02x, ",
                    param_var->byte_array_value->values[i]);
                }
              }
```

romainreignier replied:
Yes, I agree, but it is done like this for arrays of bool, integer, double and char (string).
Should I change it for the other types?
In case `values` is `NULL`, we just don't print anything?

Barry-Xu-2018 replied:
> Should I change it for the other types?

I think these issues should be addressed in a separate PR rather than being included in your PR.

> In case values is NULL, we just don't print anything?

Yes. I think it is already incorrect when `values` is NULL, and an error should have already been reported elsewhere in the code.

Barry-Xu-2018 commented on rcl_yaml_param_parser/test/benchmark/benchmark_variant.cpp:
```diff
@@ -121,6 +121,36 @@ BENCHMARK_F(PerformanceTest, string_copy_variant)(benchmark::State & st)
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+  using ArrayT = std::remove_pointer<decltype(src_variant.byte_array_value)>::type;
+  src_variant.byte_array_value =
+    static_cast<ArrayT *>(allocator.allocate(sizeof(ArrayT), allocator.state));
```
if (src_variant.byte_array_value == nullptr) {
  st.SkipWithError("Failed to allocate byte_array_value");
  return;
}

Barry-Xu-2018 commented on rcl_yaml_param_parser/test/benchmark/benchmark_variant.cpp:
```diff
@@ -121,6 +121,36 @@ BENCHMARK_F(PerformanceTest, string_copy_variant)(benchmark::State & st)
+  using ValueT = std::remove_pointer<decltype(src_variant.byte_array_value->values)>::type;
+  src_variant.byte_array_value->values =
+    static_cast<ValueT *>(
+    allocator.zero_allocate(kSize, sizeof(ValueT), allocator.state));
```
if (src_variant.byte_array_value->values == nullptr) {
  allocator.deallocate(src_variant.byte_array_value, allocator.state);
  st.SkipWithError("Failed to allocate byte_array_value->values");
  return;
}

romainreignier commented:
Thanks for the review @Barry-Xu-2018 
I have addressed your comments except the check in the loop, see my comment.

romainreignier commented:
Note, the CI fails because it does not use a version of rcutils with base64.

Barry-Xu-2018 commented on rcl_yaml_param_parser/src/parse.c:
```diff
@@ -190,7 +212,16 @@ void * get_value(
-    /// YAML_NULL_TAG, YAML_TIMESTAMP_TAG and "tag:yaml.org,2002:binary" are not supported
+    /// Check for binary tag
+    if (strcmp(BINARY_TAG, (char *)tag) == 0) {
+      if (_get_binary_array(value, val_type, &ret_val, allocator) != RCUTILS_RET_ERROR) {
```
_get_binary_array() can return RCUTILS_RET_BAD_ALLOC.

```suggestion
      if (_get_binary_array(value, val_type, &ret_val, allocator) == RCUTILS_RET_OK) {
```

Barry-Xu-2018 commented on rcl_yaml_param_parser/src/parse.c:
```diff
@@ -137,6 +142,23 @@ _get_float_value(
+/// \param[out] val_type the value type
+/// \param[out] ret_val the converted rcl_byte_array_t when value is valid
+/// \param[in] allocator the allocator to use
+/// \return RCUTILS_RET_OK if value is valid, or
```
RCUTILS_RET_BAD_ALLOC should be included.

Barry-Xu-2018 commented on rcl_yaml_param_parser/src/parse.c:
```diff
@@ -456,6 +487,31 @@ rcutils_ret_t parse_value(
       break;
+    case DATA_TYPE_BYTE:
+      if (is_seq) {
+        RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+        "Byte array must be encoded as a base64 string, not a sequence, "
+        "cannot parse value %s at line %d\n",
+        value, line_num);
+        ret = RCUTILS_RET_ERROR;
```
There will be a memory leak here because the returned `ret_val` is not handled.

Barry-Xu-2018 commented on rcl_yaml_param_parser/test/test_parse.cpp:
```diff
@@ -116,6 +116,39 @@ TEST(TestParse, parse_value) {
+
+  // byte value
+  yaml_char_t byte_value[] = "AQID";  // [0x01, 0x02, 0x03]
+  const size_t byte_value_length = sizeof(byte_value) / sizeof(byte_value[0]);
```
Here, the calculation implicitly includes the ending '\0'. Would using strlen make it clearer?
```suggestion
  const size_t byte_value_length = strlen(byte_value）
```

Barry-Xu-2018 commented on rcl_yaml_param_parser/test/test_parse.cpp:
```diff
@@ -281,6 +314,21 @@ TEST(TestParse, parse_value_sequence) {
+
+  // byte array value cannot be parsed, byte array is encoded as a base64 string, not a sequence
+  yaml_char_t byte_value[] = "AQID";  // [0x01, 0x02, 0x03]
+  const size_t byte_value_length = sizeof(byte_value) / sizeof(byte_value[0]);
```
```suggestion
  const size_t byte_value_length = strlen(byte_value);
```

Barry-Xu-2018 commented on rcl_yaml_param_parser/src/parse.c:
```diff
@@ -661,6 +717,45 @@ _get_bool_value(
+    }
+    return RCUTILS_RET_BAD_ALLOC;
+  }
+
```
These codes should be removed.
```suggestion
// Transfer ownership of the decoded buffer to dst without extra allocation.
```

Barry-Xu-2018 reviewed (commented):
Some more comments.

romainreignier commented:
Thanks a lot @Barry-Xu-2018 for your patience and careful review.
I am sorry about these errors, I am not used to code in C and all its manual memory management :/

Barry-Xu-2018 commented:
It’s totally fine :smile: 
Thank you for your contribution. I have no further comments.  
I will run CI. 
@fujitatomoya Please review this PR.

Barry-Xu-2018 commented:
Pulls: https://github.com/ros2/rcl/pull/1256
Gist: https://gist.githubusercontent.com/Barry-Xu-2018/b7bf6ed7353257611b68a0995631452b/raw/68c6270f70800a5aa34559ebe0235c15459bd997/ros2_rolling.repos
BUILD args: --packages-above-and-dependencies rcl
TEST args: --packages-above rcl
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/18147
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=27599)](http://ci.ros2.org/job/ci_linux/27599/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=20664)](http://ci.ros2.org/job/ci_linux-aarch64/20664/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=7478)](http://ci.ros2.org/job/ci_linux-rhel/7478/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=26620)](http://ci.ros2.org/job/ci_windows/26620/)

Barry-Xu-2018 commented:
All the failures were caused by errors while compiling the urdfdom_headers package.  
This failure was likely due to this newly merged commit:  
https://github.com/ros/urdfdom_headers/commit/6552c9d987ff46b2abbc2c7eb12dfce5d04a0924  
This issue had been reported at https://github.com/ros/urdfdom_headers/issues/96.  
I will use an older version of urdfdom_headers to run CI in order to avoid unrelated impacts.

Barry-Xu-2018 commented:
Pulls: https://github.com/ros2/rcl/pull/1256
Gist: https://gist.githubusercontent.com/Barry-Xu-2018/b7bf6ed7353257611b68a0995631452b/raw/68c6270f70800a5aa34559ebe0235c15459bd997/ros2_rolling.repos (Use old version 2.1.0 for urdfdom_headers)
BUILD args: --packages-above-and-dependencies rcl
TEST args: --packages-above rcl
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/18148
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=27600)](http://ci.ros2.org/job/ci_linux/27600/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=20665)](http://ci.ros2.org/job/ci_linux-aarch64/20665/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=7479)](http://ci.ros2.org/job/ci_linux-rhel/7479/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=26629)](http://ci.ros2.org/job/ci_windows/26629/)

fujitatomoya commented:
> Pulls: #1256 Gist: https://gist.githubusercontent.com/Barry-Xu-2018/b7bf6ed7353257611b68a0995631452b/raw/68c6270f70800a5aa34559ebe0235c15459bd997/ros2_rolling.repos (Use old version 2.1.0 for urdfdom_headers) BUILD args: --packages-above-and-dependencies rcl TEST args: --packages-above rcl ROS Distro: rolling Job: ci_launcher ci_launcher ran: https://ci.ros2.org/job/ci_launcher/18148
> 
> * Linux [![Build Status](https://camo.githubusercontent.com/2f2cb8235106bc3a855c6385df65f0a89806a8aeffbbb86993342ff42bd71e1c/687474703a2f2f63692e726f73322e6f72672f6275696c645374617475732f69636f6e3f6a6f623d63695f6c696e7578266275696c643d3237363030)](http://ci.ros2.org/job/ci_linux/27600/)
> * Linux-aarch64 [![Build Status](https://camo.githubusercontent.com/db70ca1e64744e26ea46450844e94628df0ea5ee7a8d9b2719afc8307b855ca7/687474703a2f2f63692e726f73322e6f72672f6275696c645374617475732f69636f6e3f6a6f623d63695f6c696e75782d61617263683634266275696c643d3230363635)](http://ci.ros2.org/job/ci_linux-aarch64/20665/)
> * Linux-rhel [![Build Status](https://camo.githubusercontent.com/4c71c379ddad9485f39d714c28cc1739137a7e94c9b1e79d02ef9d0cf3a54c78/687474703a2f2f63692e726f73322e6f72672f6275696c645374617475732f69636f6e3f6a6f623d63695f6c696e75782d7268656c266275696c643d37343739)](http://ci.ros2.org/job/ci_linux-rhel/7479/)
> * Windows [![Build Status](https://camo.githubusercontent.com/e626204351f3e6c94383c3ec33b9d43a26fbdf5f2d60f59b8cfedebe193ac574/687474703a2f2f63692e726f73322e6f72672f6275696c645374617475732f69636f6e3f6a6f623d63695f77696e646f7773266275696c643d3236363239)](http://ci.ros2.org/job/ci_windows/26629/)

@Barry-Xu-2018 CI should have been run with `rcl_yaml_param_parser`, not with `rcl`?

fujitatomoya commented:
Pulls: ros2/rcl#1256
Gist: https://gist.githubusercontent.com/fujitatomoya/8dfc89897a0d6c6ac1f30d906c4aa260/raw/905e68cf737e96ef484ff14baf120e3f41086046/ros2.repos
BUILD args:  --packages-above-and-dependencies rcl_yaml_param_parser
TEST args:  --packages-above rcl_yaml_param_parser
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/18159
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=27614)](http://ci.ros2.org/job/ci_linux/27614/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=20676)](http://ci.ros2.org/job/ci_linux-aarch64/20676/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=7490)](http://ci.ros2.org/job/ci_linux-rhel/7490/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=26633)](http://ci.ros2.org/job/ci_windows/26633/)

Barry-Xu-2018 commented:
@fujitatomoya 

> CI should have been run with rcl_yaml_param_parser, not with rcl?

Yes.  Build configuration was wrong.  Thank you for your reminder.

Barry-Xu-2018 commented:
- Linux
    - The failure was caused by CI system.
    - Re-run  [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=27616)](http://ci.ros2.org/job/ci_linux/27616/)

- Windows
The failures, `projectroot.test_action_communication__rmw_zenoh_cpp` and `rcl_action.test_action_communication__rmw_zenoh_cpp.gtest.missing_result`, are unrelated to this PR.
